src/box/watcher.c

```diff
@@ -413,6 +414,25 @@ box_broadcast(const char *key, size_t key_len,
 	watchable_broadcast(&box_watchable, key, key_len, data, data_end);
 }
 
+void
+box_broadcast_fmt(const char *key, const char *format, ...)
+{
+	struct region *gc = &fiber()->gc;
```

Comment:
I wouldn't bother with this really. You could do even simpler:
```C
void
box_broadcast_shutdown()
{
    const char *key = "box.shutdown";
    /* = mp_encode_bool(true) */
    box_broadcast(key, strlen(key), "0xc3", 1);
}
```
But if you want a more generic function, then we need to properly handle `region_alloc` and `mp_vformat` failure. The format needs to be something release-build-working (`panic()`). And `region_alloc` - either also `panic()` (we can't fail gracefully here if it would be used for an important event like leader change) or just make a local buffer on the stack + panic the body size fits into it. Should simplify it greatly:
```C
const size_t cap = 1024;
char buf[cap];
size_t size = mp_vformat(buf, cap, format, ap);
if (size >= cap)
    panic("...");
box_broadcast(key, strlen(key), buf, buf + size);
```
All up to you, I am fine with anything.

Reply:
I prefer to have a convenience helper `box_broadcast_fmt` that would allow one to easily broadcast any event using a predefined format rather than adding a new function for each event type. Event broadcast is typically a slow path so it should be fine.

I fixed `box_broadcast_fmt` accroding to your comment - I use `tt_static_buf` to avoid allocations. Note, I still use `unreachable()` for `mp_vformat()`, because it must always return the same value for the same format and arguments.

<details><summary>diff</summary>

```diff
diff --git a/src/box/watcher.c b/src/box/watcher.c
index a7b51d67ee09..2c7b378a7b64 100644
--- a/src/box/watcher.c
+++ b/src/box/watcher.c
@@ -18,6 +18,7 @@
 #include "msgpuck.h"
 #include "small/rlist.h"
 #include "trivia/util.h"
+#include "tt_static.h"
 
 /**
  * Global watchable object used by box.
@@ -417,20 +418,18 @@ box_broadcast(const char *key, size_t key_len,
 void
 box_broadcast_fmt(const char *key, const char *format, ...)
 {
-	struct region *gc = &fiber()->gc;
-	size_t svp = region_used(gc);
 	va_list ap;
 	va_start(ap, format);
 	size_t size = mp_vformat(NULL, 0, format, ap);
 	va_end(ap);
-	char *data = region_alloc(gc, size);
-	assert(data != NULL);
+	if (size > TT_STATIC_BUF_LEN)
+		panic("not enough buffer space to broadcast '%s'", key);
+	char *data = tt_static_buf();
 	va_start(ap, format);
 	if (mp_vformat(data, size, format, ap) != size)
 		unreachable();
 	va_end(ap);
 	box_broadcast(key, strlen(key), data, data + size);
-	region_truncate(gc, svp);
 }
 
 void
```

</details>